src/v/cluster/tests/eviction_stm_test.cc

```diff
@@ -0,0 +1,137 @@
+FIXTURE_TEST(test_eviction_stm_deadlock, raft_test_fixture) {
+    raft_group gr = raft_group(raft::group_id(0), 3);
+    gr.enable_all();
+    auto leader_id = wait_for_group_leader(gr);
+    auto leader_raft = gr.get_member(leader_id).consensus;
+
+    /// Publish records and open new segments
+    auto* impl = dynamic_cast<storage::disk_log_impl*>(
+      leader_raft->log().get_impl());
+    std::vector<storage::offset_stats> offsets;
+    for (auto i = 0; i < 5; ++i) {
+        replicate_random_batches(gr, 20).get0();
+        offsets.push_back(impl->offsets());
+        impl->force_roll(ss::default_priority_class()).get0();
+    }
+
+    /// Create an instance of the log eviction stm, however using the subclass
+    /// above which allows this test to drive the storage eviction loop
+    ss::abort_source as;
+    test_log_eviction_stm eviction_stm(
+      leader_raft.get(),
+      logger,
+      as,
+      gr.get_member(leader_id).storage.local().kvs());
+    eviction_stm.start().get();
+    auto cleanup = ss::defer([&] {
+        // NOTE: this it the natural order of shutdown.
+        leader_raft->stop().get();
+        as.request_abort();
+        eviction_stm.stop().get();
+    });
+
+    /// Fufills the promise causing the monitor_log_eviction loop to continue
+    const auto highest_term0_offset = offsets[0].dirty_offset;
+    eviction_stm.drive_eviction_loop(highest_term0_offset);
+    auto next_start_offset = highest_term0_offset + model::offset(1);
+    /// Wait until the effect has proceeded, i.e. a snapshot has been taken
+    tests::cooperative_spin_wait_with_timeout(
+      5s,
+      [leader_raft, highest_term0_offset] {
+          return leader_raft->last_snapshot_index() == highest_term0_offset;
+      })
+      .get();
+    /// eviction_stm should report the correct new start offset
+    BOOST_CHECK_EQUAL(next_start_offset, eviction_stm.effective_start_offset());
+
+    /// Test deadlock, attempt to truncate at offset that will be translated to
+    /// be below the raft last snapshot, this should have no effect
+    eviction_stm.drive_eviction_loop(next_start_offset);
+
+    /// Then try to evict some more data, if this step fails, previous call had
+    /// initiated the deadlock
+    const auto highest_term1_offset = offsets[1].dirty_offset;
+    eviction_stm.drive_eviction_loop(highest_term1_offset);
+    /// Wait until the effect has proceeded, i.e. a snapshot has been taken,
+    /// if this does not timeout, this means no deadlock has occurred
+    tests::cooperative_spin_wait_with_timeout(
+      5s,
+      [leader_raft, highest_term1_offset] {
+          return leader_raft->last_snapshot_index() >= highest_term1_offset;
+      })
+      .get();
+
+    next_start_offset = highest_term1_offset + model::offset(1);
+    BOOST_CHECK_EQUAL(next_start_offset, eviction_stm.effective_start_offset());
+    cleanup.cancel();
+
+    /// Shutdown the eviction stm, the abort source is externally managed
+    as.request_abort();
```

Comment:
How does this portion work with the cleanup lambda? Maybe we can just remove this or integrate the missing bit into cleanup

Reply:
The cleanup is canceled at L127, so it just doesn't run.

I initially added the cleanup lambda because it was crashing in a funny way (i botched a cherry-pick). Shouldn't be needed now, but I think it's a reasonable practice.